---
title: "Nemhauser Analysis 002"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---


## Setup the R environment

```{r setup}
library(RColorBrewer)
library(edgeR)
library(ggplot2)
library(reshape2)
library(statmod)
library(stringr)
library(VennDiagram)


# setwd('/bigdata3/projects/nem/nem.2/counts')
knitr::opts_knit$set(root.dir='/bigdata3/projects/nem/nem.2/counts')

output.dir = "/bigdata3/projects/nem/nem.2/dge.trunc"
if (! dir.exists(output.dir)) dir.create(output.dir)

sample.names <- c("3943_1", "3943_2", "3943_3", "4578_1", "4578_2", "4578_3", "4582_1", "4582_2", "4582_3")
order <- sample.names

genotype <- substr(sample.names, 1, 4)
plant <- substr(sample.names, 6, 7)

rep <- plant
group <- genotype

rep
group
```


## Define Functions

```{r functions}
customMDSplot <- function(y, dim1=1, dim2=2, mds_description=None) {
  MDS.df <- as.data.frame(plotMDS(y, dim.plot = c(dim1, dim2))$cmdscale.out[, c(dim1, dim2)])
  colnames(MDS.df) <- c('x', 'y')
  MDS.df$Genotype <- genotype
  MDS.df$Plant <- plant
  p <- ggplot(MDS.df, aes(x=x, y=y)) +
   xlab(paste("Dimension", dim1)) + ylab(paste("Dimension", dim2)) +
   ggtitle(expression("MDS Plot (Leading " * Log[2] * " fold change)")) +
   geom_point(aes(shape=Plant, color=Genotype), size=4)
  p
}
```

## Import and process counts file

```{r}
counts.file <- "counts.txt"
counts <- read.table(counts.file, sep="\t", row.names=1, as.is=T)

colnames(counts) <- sample.names
y <- DGEList(counts, genes=rownames(counts), group=group)

# Remove Meta tags
MetaTags <- grep("^__", rownames(y))
y <- y[-MetaTags, ]

y.all <- y
y.all <- calcNormFactors(y.all)


#######################################
# REMOVE THIS BLOCK TO KEEP ALL GENES #
#######################################
min.presence <- 3
min.cpm <- 1
keep <- rowSums(cpm(y) > min.cpm) >= min.presence
y <- y[keep, , keep.lib.sizes=FALSE]
#######################################
#######################################


y <- calcNormFactors(y)

design <- model.matrix(~genotype)
colnames(design)
```


## MDS Plots

```{r}
dim1 <- 1
dim2 <- 2
customMDSplot(y, dim1=dim1, dim2=dim2)
ggsave(paste0(output.dir, "/MDS.", dim1, "v", dim2, ".png"), width=7.5, height=6, units="in")

dim1 <- 3
dim2 <- 4
customMDSplot(y, dim1=dim1, dim2=dim2)
ggsave(paste0(output.dir, "/MDS.", dim1, "v", dim2, ".png"), width=7.5, height=6, units="in")
```



```{r functions.contd}
getSigGenes <- function(qlf, direction, logFC.cutoff = 0, n=Inf, logFC="logFC") {
  tt <- topTags(qlf, n=Inf)
  fc.df <- data.frame(gene.ID=rownames(tt$table), tt$table)

  de.sig <- subset(fc.df, FDR < 0.05)
  if (direction == 'up') {
    de.sig <- de.sig[de.sig[, logFC] > logFC.cutoff, ]
  } else if (direction == 'down') {
    de.sig <- de.sig[de.sig[, logFC] < -logFC.cutoff, ]
  } else if (direction == 'all') {
    de.sig <- de.sig[abs(de.sig[, logFC]) > logFC.cutoff, ]
  } else {
    stop(paste0("Invalid direction: Use 'up', 'down', or 'all' instead of '", direction, "'."))
  }
  de.sig <- de.sig[order(de.sig$FDR), ]

  head(de.sig$gene.ID, n=n)
}


plotDE <- function(genes, description="", .cpm.df=cpm.df) {
  require(reshape2)
  require(ggplot2)

  cpm.genes <- subset(.cpm.df, gene.ID %in% genes)
  m.cpm.genes <- melt(cpm.genes, variable.name="sample", value.name="cpm", id.vars=c("gene.ID"))
  m.cpm.genes$Genotype <- str_split_fixed(m.cpm.genes$sample, "_", 2)[, 1]

  p <- ggplot(m.cpm.genes, aes(x=Genotype, y=cpm)) + geom_boxplot() + facet_wrap(~gene.ID, scales="free_y") + ggtitle(description)
  p
}

plotTopDE <- function(qlf, direction, description="", n=6, logFC="logFC", .cpm.df=cpm.df) {
  require(reshape2)
  require(ggplot2)
  tt <- topTags(qlf, n=Inf)
  if (direction == "up") {
    genes <- rownames(head(tt$table[tt$table[, logFC] > 0 & tt$table[, "FDR"] < 0.05, ]
, n))
  } else if (direction == "down") {
    genes <- rownames(head(tt$table[tt$table[, logFC] < 0 & tt$table[, "FDR"] < 0.05, ]
, n))
  } else {
    stop(paste0("Invalid direction: Use 'up' or 'down' instead of '", direction, "'."))
  }
  cpm.genes <- subset(.cpm.df, gene.ID %in% genes)
  m.cpm.genes <- melt(cpm.genes, variable.name="sample", value.name="cpm", id.vars=c("gene.ID"))
  m.cpm.genes$Genotype <- str_split_fixed(m.cpm.genes$sample, "_", 2)[, 1]

  p <- ggplot(m.cpm.genes, aes(x=Genotype, y=cpm)) + geom_boxplot() + facet_wrap(~gene.ID, scales="free_y") + ggtitle(paste0(description, " (", direction, ")"))

  print(p)
  p
}


plotHeatmap <- function(qlf, description, output.dir = ".", color1="blue", color2="black", color3="red", .logcpm=logcpm) {
  tt <- topTags(qlf, n=Inf)
  genes.sig <- rownames(tt$table[tt$table[, "FDR"] < 0.05, ])
  logcpm.sig <- logcpm[rownames(logcpm) %in% genes.sig, ]
  logcpm.sig <- logcpm.sig[, order]
  hc <- hclust(as.dist(1 - cor(t(logcpm.sig))), method="average")

  description.raw <- description
  description <- gsub("[- ]", "_", description)

  heatmap(logcpm.sig, Colv=NA, Rowv=as.dendrogram(hc), labRow="", col = colorRampPalette(c(color1, color2, color3))(n = 299), main=description.raw)
  pdf(file=paste(output.dir, paste(description, "heatmap", color1, "pdf", sep="."), sep="/"), height = 7.5, width = 7.5)
  heatmap(logcpm.sig, Colv=NA, Rowv=as.dendrogram(hc), labRow="", col = colorRampPalette(c(color1, color2, color3))(n = 299), main=description.raw)
  dev.off()
}


getDEGenes <- function(qlf, description, output.dir = ".", logFC="logFC", .cpm.df=cpm.df) {
  if (! dir.exists(output.dir)) dir.create(output.dir)

  tt <- topTags(qlf, n=Inf)
  fc.df <- data.frame(gene.ID=rownames(tt$table), tt$table)

  description.raw <- description
  description <- gsub("[-]", "_", description)

  file.prefix <- paste(output.dir, description, sep="/")
  file.prefix <- gsub("[ ]", "_", file.prefix)

  de.sig <- subset(fc.df, FDR < 0.05)

  for (direction in c("down", "up")) {

    if (direction == "up") {
      de <- de.sig[de.sig[, logFC] > 0, ]
    } else {
      de <- de.sig[de.sig[, logFC] < 0, ]
    }
    de <- merge(de, .cpm.df, by = 'gene.ID', all.x=TRUE)
    de <- de[order(de$FDR), ]
    if (nrow(de) > 0) {
      write.table(de, file = paste(file.prefix, direction, "txt", sep="."), quote = FALSE, sep = "\t", row.names = FALSE)
      plotTopDE(qlf, direction, description=description.raw, logFC=logFC)
      ggsave(paste(file.prefix, direction, "top-sig", "png", sep="."), width=7.5, height=6, units="in")
    }
  }
  if (nrow(de.sig) >= 5) {
    plotHeatmap(qlf, description=description.raw, output.dir=output.dir, color1="green")
    plotHeatmap(qlf, description=description.raw, output.dir=output.dir, color1="blue")
  }
}
```


## Get Annotations

```{r annotations}
annotations <- read.delim("/data/references/arabidopsis_thaliana/genome/Arabidopsis_thaliana.TAIR10.50.annotations.txt", header = FALSE, sep = "\t")
colnames(annotations) <- c("gene.ID", "type", "name", "description")
head(annotations)

annotations$summary <- paste(annotations$gene.ID, annotations$name, annotations$description, sep = "|")
annotations$summary <- gsub("\\|+", "\\|", annotations$summary)
annotations$summary <- gsub("\\|+$", "", annotations$summary)

head(annotations)
tail(annotations)
```


## Process and save cpm file

```{r}
cpm <- cpm(y)
cpm <- cpm[, order]
cpm.df <- data.frame(gene.ID=rownames(cpm), cpm, stringsAsFactors=FALSE, check.names = FALSE)
cpm.df.for.file <- merge(cpm.df, annotations, by = "gene.ID", all.x=TRUE)
cpm.df.for.file[is.na(cpm.df.for.file)] <- ""
head(cpm.df.for.file)

cpm.filename <- paste(output.dir, "cpm.txt", sep="/")
write.table(cpm.df.for.file, cpm.filename, sep="\t", quote=F, row.names=F)

logcpm <- cpm(y, prior.count=2, log=TRUE)
```


## Differential Gene Expression

```{r}
y <- estimateDisp(y, design)
fit <- glmFit(y, design)
```


#### Genes responding to the presence of endogenous GA in 4578

```{r}
description <- "Endogenous GA-responsive in 4578"
lrt <- glmLRT(fit, coef=2)
summary(decideTestsDGE(lrt))
up.4578 <- getSigGenes(lrt, "up")
down.4578 <- getSigGenes(lrt, "down")
topTags(lrt)
getDEGenes(lrt, description, output.dir=output.dir, .cpm.df=cpm.df.for.file)
```


#### Genes responding to the presence of endogenous GA in 4582

```{r}
description <- "Endogenous GA-responsive in 4582"
lrt <- glmLRT(fit, coef=3)
summary(decideTestsDGE(lrt))
up.4582<- getSigGenes(lrt, "up")
down.4582 <- getSigGenes(lrt, "down")
topTags(lrt)
getDEGenes(lrt, description, output.dir=output.dir, .cpm.df=cpm.df.for.file)
```



```{r}
for (gene in setdiff(up.4578, up.4582)) {
  if (! dir.exists(paste(output.dir, "up.4578.only", sep="/"))) dir.create(paste(output.dir, "up.4578.only", sep="/"))
  plotDE(gene, description="Up in 4578")
  ggsave(paste0(output.dir, "/up.4578.only/", gene, ".png"), width=7.5, height=6, units="in")
}

for (gene in setdiff(down.4578, down.4582)) {
  if (! dir.exists(paste(output.dir, "down.4578.only", sep="/"))) dir.create(paste(output.dir, "down.4578.only", sep="/"))
  plotDE(gene, description="Down in 4578")
  ggsave(paste0(output.dir, "/down.4578.only/", gene, ".png"), width=7.5, height=6, units="in")
}

for (gene in setdiff(up.4582, up.4578)) {
  if (! dir.exists(paste(output.dir, "up.4582.only", sep="/"))) dir.create(paste(output.dir, "up.4582.only", sep="/"))
  plotDE(gene, description="Up in 4582")
  ggsave(paste0(output.dir, "/up.4582.only/", gene, ".png"), width=7.5, height=6, units="in")
}

for (gene in setdiff(down.4582, down.4578)) {
  if (! dir.exists(paste(output.dir, "down.4582.only", sep="/"))) dir.create(paste(output.dir, "down.4582.only", sep="/"))
  plotDE(gene, description="Down in 4582")
  ggsave(paste0(output.dir, "/down.4582.only/", gene, ".png"), width=7.5, height=6, units="in")
}

for (gene in intersect(up.4578, up.4582)) {
  if (! dir.exists(paste(output.dir, "up.4578_4582", sep="/"))) dir.create(paste(output.dir, "up.4578_4582", sep="/"))
  plotDE(gene, description="Up in 4578 & 4582")
  ggsave(paste0(output.dir, "/up.4578_4582/", gene, ".png"), width=7.5, height=6, units="in")
}

for (gene in intersect(down.4578, down.4582)) {
  if (! dir.exists(paste(output.dir, "down.4578_4582", sep="/"))) dir.create(paste(output.dir, "down.4578_4582", sep="/"))
  plotDE(gene, description="Down in 4578 & 4582")
  ggsave(paste0(output.dir, "/down.4578_4582/", gene, ".png"), width=7.5, height=6, units="in")
}

plotDE(setdiff(up.4578, up.4582), description="Up in 4578")
ggsave(paste0(output.dir, "/up.4578.png"), width=7.5, height=6, units="in")

plotDE(setdiff(down.4578, down.4582), description="Down in 4578")
ggsave(paste0(output.dir, "/down.4578.png"), width=7.5, height=6, units="in")

plotDE(setdiff(up.4582, up.4578), description="Up in 4582")
ggsave(paste0(output.dir, "/up.4582.png"), width=7.5, height=6, units="in")

plotDE(setdiff(down.4582, down.4578), description="Down in 4582")
ggsave(paste0(output.dir, "/down.4582.png"), width=7.5, height=6, units="in")

plotDE(intersect(up.4578, up.4582), description="Up in 4578 & 4582")
ggsave(paste0(output.dir, "/up.4578_4582.png"), width=7.5, height=6, units="in")

plotDE(intersect(down.4578, down.4582), description="Down in 4578 & 4582")
ggsave(paste0(output.dir, "/down.4578_4582.png"), width=7.5, height=6, units="in")
```

#### Venn Diagram of DE Genes

```{r}
fourway.Venn <- function(A, B, C, D, cat.names = c("A", "B", "C", "D")){
  area1 <- length(A)
  area2 <- length(B)
  area3 <- length(C)
  area4 <- length(D)
  n12 <- length(intersect(A, B))
  n13 <- length(intersect(A, C))
  n14 <- length(intersect(A, D))
  n23 <- length(intersect(B, C))
  n24 <- length(intersect(B, D))
  n34 <- length(intersect(C, D))
  n123 <- length(intersect(A, intersect(B, C)))
  n124 <- length(intersect(A, intersect(B, D)))
  n134 <- length(intersect(A, intersect(C, D)))
  n234 <- length(intersect(B, intersect(C, D)))
  n1234 <- length(intersect(A, intersect(B, intersect(C, D))))
  venn.plot <- draw.quad.venn(
    area1 = area1,
    area2 = area2,
    area3 = area3,
    area4 = area4,
    n12 = n12,
    n13 = n13,
    n14 = n14,
    n23 = n23,
    n24 = n24,
    n34 = n34,
    n123 = n123,
    n124 = n124,
    n134 = n134,
    n234 = n234,
    n1234 = n1234,
    category = cat.names,
    fill = c(
      "orange",
      "forestgreen",
      "red",
      "blue"
    ),
    alpha = .6,
    cex = 2,
    cat.cex = 2
  )

  # Add comma separators for larger numbers (https://stackoverflow.com/a/37240111/996114)
  idx <- sapply(venn.plot, function(i) grepl("text", i$name))
  for(i in 1:7){
    venn.plot[idx][[i]]$label <- format(as.numeric(venn.plot[idx][[i]]$label), big.mark=",", scientific=FALSE)
  }
  venn.plot
}

png(paste(output.dir, 'venn.png', sep="/"), w=7, h=7, u='in', res=150)
venn.de <- fourway.Venn(
  up.4582,
  down.4582,
  up.4578,
  down.4578,
  cat.names = c(
    "Up 4582",
    "Dn 4582",
    "Up 4578",
    "Dn 4578"
  ))
grid.newpage()
grid.draw(venn.de)
dev.off()

pdf(paste(output.dir, 'venn.pdf', sep="/"), w=7, h=7, useDingbats = FALSE)
grid.draw(venn.de)
dev.off()

grid.newpage()
grid.draw(venn.de)

up.both <- sort(intersect(up.4578, up.4582))
down.both <- sort(intersect(down.4578, down.4582))
up.both
down.both

write.table(up.both, file = paste(output.dir, "up-both.txt", sep="/"), quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)
write.table(down.both, file = paste(output.dir, "down-both.txt", sep="/"), quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)
```
